andi/andi.py

```diff
@@ -53,10 +59,161 @@ def to_provide(
     if isinstance(can_provide, Container):
         can_provide = can_provide.__contains__
 
+    bindings = bindings or (lambda x: None)
     result = {}
     for argname, types in arguments.items():
-        for cls in types:
-            if can_provide(cls):
-                result[argname] = cls
-                break
+        sel_cls = select_type(types, can_provide, bindings)
+        if sel_cls:
+            result[argname] = sel_cls
     return result
+
+
+Plan = typing.Dict[Type, Dict[str, Type]]
+
+
+class NonProvidableError(TypeError):
+    """ Raised when a type is not providable """
+
+
+class CyclicDependencyError(TypeError):
+    """ Raised on cyclic dependencies """
+
+
+class FunctionArguments:
+    """ Key marker to return the inspected function arguments into the
+    ``Plan`` returned by the ``plan`` function """
+    pass
+
+
+def plan(class_or_func: Union[Type, Callable],
+         can_provide: TypeContainerOrCallable,
+         externally_provided: TypeContainerOrCallable,
+         bindings: Callable[[Type], Optional[Type]] = None) -> Plan:
+    """ Check if it is possible to fulfill the arguments to invoke the input
+    function (or to the create the input class if a class is given). If possible
+    then a plan to build the requirements is returned. This plan is
+    an ``OrderedDict`` containing the proper instantiation order
+    for each type so that the dependencies existence is assured. The keys in
+    this plan are the type, and the values are dicts where the keys are the
+    params names and the values are the type required for this parameter.
+
+    If the input is a function then there will be a final entry in the plan with
+    the key ``FunctionArguments`` that contains the dictionary of the
+    fulfilled arguments for this function (could be incomplete).
```

Comment:
Let me summarize my understanding of this. `andi.plan` can receive either class or a callable as a first argument.

Callable:

* arguments of this callable are inspected
* no exception is raised if some of them can't be satisfied, because they're not in can_provide mapping
* final entry in the plan has FunctionArguments key, and contains `{"argument_name": cls, ...}` dictionary, with all arguments which can be provided

Class:
* arguments of `cls.__init__` are inspected
* NonProvidableError is raised if some of the arguments can't be satisfied, because they're not in can_provide
* `cls.__init__` argument which are not type annotated are not inspected, and no error is raised for them, even if they don't have a default value
* final entry in the plan has `cls` key, and contains `{"argument_name": cls, ...}` dictionary, with all type annotated `cls.__init__` arguments.

Previously a design was a bit different: to_provide (which is like "flat" version of "plan"?) was not raising an error. If checking of arguments is required, user was expected to 

* call andi.inspect first, get a list of arguments which can be potentially handled, then
* call andi.to_provide, to get a list of arguments which can be handled in practice, then
* compare the results, and raise an error if needed (error is going to be raise later anyways, but here is a chance to get a better error message, or fail earlier, if we know it is not going to work)

I actually had a similar exception in initial designs of to_provide, but removed it, because getting enough flexibility was complicating code too much.

I wonder if this flexibility issue applies here as well, and if such error checking can be moved outside of `plan`function, at least for the original, top-level object. 

There are use cases for validating that all Callable arguments can be satisfied, and there can be use cases for having not all class arguments validated (though this one can be solved currently by passing `cls.__init__` explicitly).

I hope that it should allow to 

1) implement "page object functions" more easily in scrapy-po, and give more flexibility overall: don't dictate that "for functions we don't validate, for classes we do validate";
2) simplify documentation for this new feature (it took me some time to understand what would be the difference between passing a class and passing a callable); remove an implicit convention in favor of either an additional function call, or additional function argument, or something else
3) remove FunctionArguments key, which needs to be handled in a condition in the building code.

As a final entry in the Plan is special, I wonder if it makes sense to make Plan a class, which provides access to the plan itself, and to the last step separately. This may allow to remove a condition from the building code (FunctionArguments check in a loop); printing code can also be moved to a method.

Reply:
TL;DR: I will work on your feedback. Now the full description:

> cls.__init__ argument which are not type annotated are not inspected, and no error is raised for them, even if they don't have a default value

Great catch. This is something I completely missed. It should fail for non-annotated arguments without a default value. Or even more strictly, it shouldn't be allowed parameters without type annotations. Allowing non-annotated arguments with default values could be useful for testing (i.e. `debug=False`) so probably better to allow them. 

----
I understand your concerns. Let me first clarify something. As you noted the function plan is mixing two use cases (when a function is given or when a class is given) that are completely different and this could be messing the things up (difficult to explain the behaviour, difficult documentation). I was experiment with having two methods, one for the function and one for the class, but I was not happy with the result. But I think is it worth to try it again now that I have things more clear. The two use cases are the following:
* **Case function (top level):**: You have a function with arguments, and you may want to inject some of them. It can be right not being able to resolve are dependencies because you can resolve them maybe by other means (i.e. `response` from Scrapy). 
* **Case class (recursion):** You need to build a class and you need **all dependencies to be resolved** because you need to be able to build the class without any other external help. And you need that because this class is a requirement appearing in the tree of dependencies for the first case (the top level). In other words, you may not require being able to resolve all dependencies of a function, but you definitely need that all dependencies you told could be resolved can actually be built. 

There is an alternative behaviour for this second case that is that if a class cannot be built, instead of raising an exception, just don't return the dependant top level attribute as possible in the "case function". I don't like this as default case because most likely it is a failure and you can have a better message if failing early, but we could have a flag to enable/disable this behaviour so that the caller decide what to do. 

Note that nothing prevents to use the "case function" for building classes by calling `andi.plan(TheClass.__init__,...)`

----
Comments to your hopes:

> implement "page object functions" more easily in scrapy-po, and give more flexibility overall: don't dictate that "for functions we don't validate, for classes we do validate";

Still, I think it is mandatory for internal classes to be "buildable" and otherwise fail. Do you any case where it would be useful a different behaviour? If this is not forced either the plan would be incomplete (so not possible to follow, bad idea) or we just return the top branches of the trees that are possible but not sure when this would be useful. The former behaviour could be enabled/disabled using a flag as I told before. 

> simplify documentation for this new feature (it took me some time to understand what would be the difference between passing a class and passing a callable); remove an implicit convention in favor of either an additional function call, or additional function argument, or something else

Sure, let me play a little bit with that. 

> remove FunctionArguments key, which needs to be handled in a condition in the building code.

I'm not sure about that (more about it below), but let me play with it as well. 

> Plan a class, which provides access to the plan itself, and to the last step separately.

I also experimented with that, but having the last item of the plan separated is not useful, because you always need to follow the plan in sequence and this would force to have special build logic for the last case, even more special than the one required one with the `FunctionArguments` check. But I'll have a second thought on it.

Reply:
Hey! Regarding case classes vs case functions: I agree that creation should fail if an internal dependency can't be satisfied completely. There can be cases where we don't want it, but they're very complicated - e.g. separate components are updating the plan, and so individual calls shouldn't fail if something can't be satisfied; validation should run only after all "plan iterations" are done. Let's not think about it :)

I guess my main point is that distinction is not actually between classes and functions, but it is between objects for which we want to satisfy arguments completely, and objects where it is OK to satisfy arguments partially. There can be cases where we want to ensure all function arguments are satisfied, and cases where that's fine to skip some of the class init arguments; because of that for me an explicit flag, or moving validation externally, looks better than relying on class/function distinction.

Reply:
> It can be right not being able to resolve are dependencies because you can resolve them maybe by other means (i.e. response from Scrapy).

Another example is having several packages / middlewares doing dependency injection in Scrapy. For example, there can be a middleware which passes CookieJar if a callback receives an argument annotated as CookieJar, and it can be in a separate package (or a built-in Scrapy middleware), not related to scrapy-po.

Reply:
> cls.__init__ argument which are not type annotated are not inspected, and no error is raised for them, even if they don't have a default value

I fixed that here https://github.com/scrapinghub/andi/pull/8/commits/1773cbb9f9accd21d529a4a6079285956a4b700a in the strict way: failing even if they have a default value. Let's cover the case with defaults in the future, because it requires changing `inspect` API to also return the defaults. 

> There can be cases where we want to ensure all function arguments are satisfied, and cases where that's fine to skip some of the class init arguments; because of that for me an explicit flag, or moving validation externally, looks better than relying on class/function distinction.

I'm going to work on that now.